### PR TITLE
Detect changes and type fixes

### PR DESCRIPTION
**Pull request by shahnami**
This PR addresses https://github.com/OpenZeppelin/defender-serverless/issues/25 and https://github.com/OpenZeppelin/defender-serverless/issues/26.

- Fixed some typings and schema validation
- Detect changes before calling 'update' (**caveat**: this does not happen for secrets, as we cannot retrieve values to compare).

**Check list:** 

- [x] https://github.com/OpenZeppelin/defender-client/pull/166

### CODE REVIEW

**Review by mverzilli — APPROVED** [review]

File: src/cmd/deploy.ts [hunk 1]
```diff
@@ -507,20 +546,94 @@ export default class DefenderDeploy {
         retrieveExisting,
         // on update
         async (sentinel: YSentinel, match: DefenderSentinel) => {
+          const isForta = (o: DefenderSentinel): o is DefenderFortaSentinelResponse => o.type === 'FORTA';
+          const isBlock = (o: DefenderSentinel): o is DefenderBlockSentinelResponse => o.type === 'BLOCK';
+
+          // Warn users when they try to change the sentinel network
+          if (match.network !== sentinel.network) {
+            this.log.warn(
+              `Detected a network change from ${match.network} to ${sentinel.network} for Sentinel: ${match.stackResourceId}. Defender does not currently allow updates to the network once a Sentinel is created. This change will be ignored. To enforce this change, remove this sentinel and create a new one. Atlernatively, you can change the unique identifier (stack resource ID), to force a new creation of the sentinel.`,
```
Comment (spalladino) [inline src/cmd/deploy.ts:555]: Does this apply to all sentinels, or only BLOCK ones?

File: src/utils/index.ts [hunk 2]
```diff
@@ -25,6 +25,7 @@ import {
   ResourceType,
   DefenderBlockWatcher,
 } from '../types';
+import { sanitize } from './sanitise';
```
Comment (spalladino) [inline src/utils/index.ts:28]: :us: :crossed_swords: :gb:

File: src/utils/sanitise.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+const _sanitizeArray = (array: Array<any>) => {
+  return _.filter(_.map(array, sanitize), _isProvided);
+};
+
+const _sanitizeObject = (object: object) => {
+  return _.pickBy(_.mapValues(object, sanitize), _isProvided);
+};
+
+const _isProvided = (value: any) => {
+  const typeIsNotSupported = !_.isNull(value) && !_.isString(value) && !_.isArray(value) && !_.isPlainObject(value);
```
Comment (spalladino) [inline src/utils/sanitise.ts:27]: Hold on, aren't numbers or booleans supported?

File: src/utils/index.ts [hunk 4]
```diff
@@ -37,11 +38,16 @@ export const getEquivalentResource = <Y, D>(
...
   if (resource) {
-    const [key, _] = Object.entries(resources ?? []).find((a) => isEqual(a[1], resource))!;
+    const [key, value] = Object.entries(resources ?? []).find((a) => _.isEqual(a[1], resource))!;
     return currentResources.find((e: D) => (e as any).stackResourceId === getResourceID(getStackName(context), key));
   }
 };
 
+export const removeNils = (o: object): object => {
+  // The conversion is a neat hack that removes nested `undefined` values
+  return JSON.parse(JSON.stringify(sanitize(o)));
```
Comment (spalladino) [inline src/utils/index.ts:48]: Do you really need the parse/stringify, given sanitize is already recursively going through the data structure and removing anything empty-ish?

File: src/utils/index.ts [hunk 5]
```diff
@@ -37,11 +38,16 @@ export const getEquivalentResource = <Y, D>(
   currentResources: D[],
 ) => {
   if (resource) {
-    const [key, _] = Object.entries(resources ?? []).find((a) => isEqual(a[1], resource))!;
+    const [key, value] = Object.entries(resources ?? []).find((a) => _.isEqual(a[1], resource))!;
     return currentResources.find((e: D) => (e as any).stackResourceId === getResourceID(getStackName(context), key));
   }
 };
 
+export const removeNils = (o: object): object => {
```
Comment (spalladino) [inline src/utils/index.ts:46]: Let's rename this function to be clear that it does a lot more than removing nils, it also drops any unsupported types

**Review by spalladino — COMMENTED** [review]
LGTM, except for the type checks in _isProvided, which make me a bit nervous. Also, given that we're now removing all non-empty props, can we test that the plugin properly recognizes updates when we change something from null to non-null and vice versa? Eg setting an autotask condition or trigger for a sentinel.

File: src/cmd/deploy.ts (hunk 1 above)
Comment (shahnami) [reply src/cmd/deploy.ts:555]: All, should it be only block sentinels?

File: src/utils/index.ts (hunk 4 above)
Comment (shahnami) [reply src/utils/index.ts:48]: This was my paranoia kicking in 😅

File: src/utils/sanitise.ts (hunk 3 above)
Comment (shahnami) [reply src/utils/sanitise.ts:27]: Good catch!

File: src/utils/index.ts (hunk 4 above)
Comment (spalladino) [reply src/utils/index.ts:48]: Hehe I fear that going to json and back may bring in more issues than it solves

File: src/cmd/deploy.ts (hunk 1 above)
Comment (spalladino) [reply src/cmd/deploy.ts:555]: You're right, I just tested and it's not possible to change network in FORTA sentinels. Disregard this comment!

File: src/utils/sanitise.ts (hunk 3 above)
Comment (shahnami) [reply src/utils/sanitise.ts:27]: So, after executing `sanitise` with:

```
 sanitise({
    aTextNumber: '123',
    aTrueBool: true,
    aFalseBool: false,
    aTextFalseBool: 'false',
    aTextTrueBool: 'true',
    aNumber: 123,
    aTrueNumber: 1,
    aFalseNumber: 0,
    anUndefined: undefined,
    aNull: null,
    aTextUndefined: "undefined",
    aTextNull: "null"
  }),
```

I'm getting:
```
aFalseBool: false
aFalseNumber: 0
aNumber: 123
aTextFalseBool: "false"
aTextNull: "null"
aTextNumber: "123"
aTextTrueBool: "true"
aTextUndefined: "undefined"
aTrueBool: true
aTrueNumber: 1
```

Which is expected behaviour.

`isEmpty` will yield true for any number or bool.

File: src/types/docs/function-properties-trigger.md [hunk 6]
```diff
@@ -4,17 +4,20 @@
 
 one (and only one) of
 
-*   [Cron](function-properties-trigger-oneof-cron.md "check type definition")
+*   [Untitled  type in Function](function-properties-trigger-oneof-0.md "check type definition")
 
-*   [Frequecy](function-properties-trigger-oneof-frequecy.md "check type definition")
+*   [Untitled  type in Function](function-properties-trigger-oneof-1.md "check type definition")
```
Comment (mok0230) [inline src/types/docs/function-properties-trigger.md:9]: this looks autogenerated, but these "Untitled type in Function" items look a little weird, are they missing a field or something?

File: src/types/docs/function.md [hunk 7]
```diff
@@ -71,9 +71,11 @@
 
 one (and only one) of
 
-*   [Cron](function-properties-trigger-oneof-cron.md "check type definition")
+*   [Untitled  type in Function](function-properties-trigger-oneof-0.md "check type definition")
```
Comment (mok0230) [inline src/types/docs/function.md:74]: same question as above

File: src/utils/index.ts [hunk 8]
```diff
@@ -199,7 +204,7 @@ export const constructSentinel = (
     name: sentinel.name,
     network: sentinel.network,
     addresses: sentinel.addresses,
-    abi: sentinel.abi && JSON.stringify(sentinel.abi),
+    abi: sentinel.abi && JSON.stringify(JSON.parse(sentinel.abi)),
```
Comment (mok0230) [inline src/utils/index.ts:207]: out of curiosity is `JSON.stringify(JSON.parse(a))` not equal to `a` in some cases?

File: src/utils/sanitise.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
+/**
+ * @notice copied from https://gist.github.com/simonrenoult/2fa79fec8f035b2d35cef0767363b227
+ * */
+
+import _ from 'lodash';
+
+export const sanitise = (object: object): object | string | null => {
```
Comment (mok0230) [inline src/utils/sanitise.ts:7]: Ha! Saw Palla's comment on this and felt like the spelling inconsistency would drive me to insanity. Though I prefer the "z", very content with consistency between function and file name!

**Review by mok0230 — APPROVED** [review]
a few very minor things, but looks great overall!

File: src/utils/index.ts (hunk 8 above)
Comment (spalladino) [reply src/utils/index.ts:207]: Whenever you're not dealing with basic types. Some things just fail to be serialized:
```
> JSON.parse(JSON.stringify({ bigint: BigInt(10) }))
Uncaught TypeError: Do not know how to serialize a BigInt
```

Others are ignored:
```
> JSON.parse(JSON.stringify({ f: (x) => x }))
{}
```

And others are converted into a plain object that describes them:
```
> b = Buffer.from('1234')
<Buffer 31 32 33 34>
> JSON.parse(JSON.stringify({ b }))
{ b: { type: 'Buffer', data: [ 49, 50, 51, 52 ] } }
> b.toString('base64')
'MTIzNA=='
> JSON.parse(JSON.stringify({ b })).b.toString('base64')
'[object Object]'
```

Comment (spalladino) [reply src/utils/index.ts:207]: Oops, I just noted I posted on the converse (stringify then parse, where you had asked for parse then stringify). I think the one you mention is safe, since parse should only return basic types.

File: src/types/docs/function-properties-trigger.md (hunk 6 above)
Comment (shahnami) [reply src/types/docs/function-properties-trigger.md:9]: Correct, they are missing a title field in the main schema. The problem is that adding them to oneOf (or required) attributes causes the serverless schema validation to fail. Haven't found a way around that yet. Fortunately, the files that are referenced in the Readme are fine.

File: src/utils/index.ts (hunk 8 above)
Comment (shahnami) [reply src/utils/index.ts:207]: > out of curiosity is `JSON.stringify(JSON.parse(a))` not equal to `a` in some cases?

Mainly for getting rid of \n and undefined values.

Comment (mok0230) [reply src/utils/index.ts:207]: > Oops, I just noted I posted on the converse

Very insightful regardless! Definitely learned something. Feel like this would have been a winning answer on Stack Overflow for such a question
